src/Microsoft.ML.Data/DataLoadSave/TransformerChain.cs

```diff
@@ -36,20 +36,33 @@ public enum TransformerScope
         Everything = Training | Testing | Scoring
     }
 
+    [BestFriend]
+    internal interface ITransformerAccessor
+    {
+        ITransformer[] Transformers { get; }
+        TransformerScope[] Scopes { get; }
+    }
+
     /// <summary>
     /// A chain of transformers (possibly empty) that end with a <typeparamref name="TLastTransformer"/>.
     /// For an empty chain, <typeparamref name="TLastTransformer"/> is always <see cref="ITransformer"/>.
     /// </summary>
-    public sealed class TransformerChain<TLastTransformer> : ITransformer, ICanSaveModel, IEnumerable<ITransformer>
+    public sealed class TransformerChain<TLastTransformer> : ITransformer, ICanSaveModel, IEnumerable<ITransformer>, ITransformerAccessor
     where TLastTransformer : class, ITransformer
```

Comment:
This is confusing. You now can access the inner transformers in 3 ways: via `foreach` (as an `IEnumerable`), via the internal field `Transformers` and via `ITransformerAccessor.Transformers`. Let's reduce it to one. Maybe make in an `IEnumerable` of pairs, if you need access to `scopes`

Reply:
Or you could make `Transformers` and `Scopes` public and `IReadOnlyList`s

---
In reply to: [236007659](https://github.com/dotnet/machinelearning/pull/1618#discussion_r236007659) [](ancestors = 236007659)

Reply:
I created this interface so that I can access transformers and scope from an ITransfomer reference because casting ITransfomer to TransformerChain<ITransformer> is messy.

---
In reply to: [236007746](https://github.com/dotnet/machinelearning/pull/1618#discussion_r236007746) [](ancestors = 236007746,236007659)